src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala

```diff
@@ -18,97 +19,149 @@ import scala.util.control.NonFatal
         // Used for task termination or as a result from a UpdateStatus action.
         // The expunge is propagated to the taskTracker which in turn informs the sender about the success (see Ack).
         repo.expunge(op.taskId.idString).map { _ =>
-          taskTrackerRef ! TaskTrackerActor.TaskRemoved(op.taskId, TaskTrackerActor.Ack(op.sender))
-        }.recoverWith(tryToRecover(op)(expectedTaskState = None))
-
-      case Action.UpdateStatus(status) =>
-        statusUpdateActionResolver.resolve(op.taskId, status).flatMap { action: Action =>
-          // Since this action is mapped to another action, we delegate the responsibility to inform
-          // the sender to that other action.
-          process(op.copy(action = action))
-        }
+          taskTrackerRef ! TaskTrackerActor.TaskRemoved(op.taskId, TaskTrackerActor.Ack(op.sender, change))
+        }.flatMap { _ =>
+          notifyOthers(op.stateOp, change)
+        }.recoverWith(tryToRecover(op)(expectedTaskState = None, oldTaskState = Some(change.task)))
 
-      case Action.Noop =>
+      case change: TaskStateChange.Failure =>
+        // Used if a task status update for a non-existing task is processed.
+        // Since we did not change the task state, we inform the sender directly of the failed operation.
+        op.sender ! Status.Failure(change.cause)
+        Future.successful(())
+
+      case change: TaskStateChange.NoChange =>
         // Used if a task status update does not result in any changes.
         // Since we did not change the task state, we inform the sender directly of the success of
         // the operation.
-        op.sender ! (())
+        op.sender ! change
         Future.successful(())
 
-      case Action.Fail(cause) =>
-        // Used if a task status update for a non-existing task is processed.
-        // Since we did not change the task state, we inform the sender directly of the failed operation.
-        op.sender ! Status.Failure(cause)
+      case change: TaskStateChange.Update =>
+        // Used for a create or as a result from a UpdateStatus action.
+        // The update is propagated to the taskTracker which in turn informs the sender about the success (see Ack).
+        val marathonTask = TaskSerializer.toProto(change.newState)
+        repo.store(marathonTask).map { _ =>
+          taskTrackerRef ! TaskTrackerActor.TaskUpdated(change.newState, TaskTrackerActor.Ack(op.sender, change))
+        }.flatMap { _ =>
+          notifyOthers(op.stateOp, change)
+        }.recoverWith(tryToRecover(op)(
+          expectedTaskState = Some(change.newState), oldTaskState = change.oldState))
+    }
+  }
+
+  private[this] def notifyOthers(stateOp: TaskStateOp, stateChange: TaskStateChange): Future[Unit] = {
+    // compatibility: if it's a statusUpdate that resulted in a Update or Expunge, do as before
+
+    stateChange match {
+      case _: TaskStateChange.Update | _: TaskStateChange.Expunge =>
```

Comment:
This logic is redundant. You only call the `notifyOthers` method for these cases so you don't really check for the relevant ops again.